src/inmanta/ast/statements/generator.py

```diff
@@ -93,6 +93,7 @@ def execute(self, requires: Dict[object, object], instance: Resolver, queue: Que
         Evaluate this statement
         """
         LOGGER.log(LOG_LEVEL_TRACE, "executing subconstructor for %s implement %s", self.type, self.implements.location)
+        assert isinstance(instance, Instance)
```

Comment:
I think this assertion is an indication that the typing here still isn't sufficiently advanced. I think I would prefer to leave this a type error (or move the type error to the method signature) than to cover it up with an assertion. On the other hand, an assertion will make sure our tests fail if this is ever triggered with an incorrect value. What do you think.

Reply:
Good point.

I would not attempt to fix this here, as it will require actual changes, not just some typing cleanup. This will quickly spiral out of hand. (I call upon the overriding review criterium: https://internal.inmanta.com/developer/tasks/commiting_changes.html#test-and-review)

But I do see the advantage of leaving this as an explicit type error.

Do you think I should?

Reply:
(or do both, the type error and the assertion)

Reply:
I definitely agree with the overriding review criterium, fixing this would be far from straightforward. Having both the assertion and the type error would be nice but I don't really see how you would achieve that (cleanly), do you?

Reply:
made an attempt

Reply:
Looks good to me. Perhaps add a comment above the assertion stating that it protects against consequences of a type error in the type of the parameter?